src/xmipp/libraries/reconstruction_cuda/cuda_angular_sph_alignment.cpp

```diff
@@ -0,0 +1,648 @@
+    std::vector<double> dVec(tVec.begin(), tVec.end());
+    memcpy(outputImage().data, dVec.data(), sizeof(double) * elements);
+}
+
+void AngularSphAlignment::setupVolumeData() 
+{
+    const auto& vol = program->V();
+    transformData(&dVolData, vol.data, vol.zyxdim, dVolData == nullptr);
+}
+
+void AngularSphAlignment::setupVolumeDataCpu() 
+{
+    const auto& vol = program->V();
+    volDataVec.assign(vol.data, vol.data + vol.zyxdim);
+    dVolData = volDataVec.data();
+}
+
+void AngularSphAlignment::setupRotation() 
+{
+    transformData(&dRotation, program->R.mdata, program->R.mdim, dRotation == nullptr);
+}
+
+void AngularSphAlignment::setupRotationCpu() 
+{
+    rotationVec.assign(program->R.mdata, program->R.mdata + program->R.mdim);
+    dRotation = rotationVec.data();
+}
+
+void AngularSphAlignment::setupVolumeMask() 
+{
+    if (dVolMask == nullptr) {
+        if (cudaMallocAndCopy(&dVolMask, program->V_mask.data, program->V_mask.getSize())
+                != cudaSuccess)
+            processCudaError();
+    } else {
+        if (cudaMemcpy(dVolMask, program->V_mask.data, program->V_mask.getSize() * sizeof(int),
+                    cudaMemcpyHostToDevice) != cudaSuccess)
+            processCudaError();
+    }
+}
+
+void AngularSphAlignment::setupVolumeMaskCpu()
+{
+    dVolMask = program->V_mask.data;
+}
+
+void AngularSphAlignment::setupProjectionPlane() 
+{
+    const auto& projPlane = program->P();
+    if (dProjectionPlane == nullptr) {
+        if (cudaMalloc(&dProjectionPlane, projPlane.yxdim * sizeof(PrecisionType)) != cudaSuccess)
+            processCudaError();
+    }
+    cudaMemset(dProjectionPlane, 0, projPlane.yxdim * sizeof(PrecisionType));
+}
+
+void AngularSphAlignment::setupProjectionPlaneCpu() 
+{
+    const auto& projPlane = program->P();
+    projectionPlaneVec.assign(projPlane.data, projPlane.data + projPlane.yxdim);
+    dProjectionPlane = projectionPlaneVec.data();
+}
+
+void fakeKernel(
+        PrecisionType Rmax2,
+        PrecisionType iRmax,
+        ImageMetaData volMeta,
+        PrecisionType* volData,
+        PrecisionType* rotation,
+        int steps,
+        int4* zshparams,
+        PrecisionType3* clnm,
+        int* Vmask,
+        PrecisionType* projectionPlane,
+        KernelOutputs* outputs
+        );
+
+void AngularSphAlignment::runKernel() 
+{
+    if (program->useFakeKernel) {
+        fakeKernel(
+                Rmax2,
+                iRmax,
+                imageMetaData,
+                dVolData,
+                dRotation,
+                steps,
+                dZshParams,
+                dClnm,
+                dVolMask,
+                dProjectionPlane,
+                &outputs);
+    } else {
+        // Define thrust reduction vector
+        thrust::device_vector<PrecisionType> thrustVec(totalGridSize * 3, 0.0);
+
+        // TEST make sure everything is ready before kernel starts
+        cudaDeviceSynchronize();
```

Comment:
you should use your `gpu` instance and synchronize only your stream, not whole device